### ISSUE TITLE
Problems in model adaptation

### ISSUE THREAD

**Issue by PabloMese** [ContinualAI/avalanche#1163]
Hi, I'm implementing my own code based in some of the Avalanche features and I'm experiencing some issue with the model adaptation in a class incremental setting. 

I'm following the code in [ https://avalanche-api.continualai.org/en/v0.1.0/_modules/avalanche/models/dynamic_modules.html#IncrementalClassifier](url). After a model being trained in the baseline experience and when I pass its weights and bias to start training the first incremental experience, the model forgets the previous classes due to the weights and bias of the new linear.

Anyone has experimented the same issue or has any idea to solve it?

Thank you in adavance

Pablo

**Comment by AndreaCossu** [comment]
If you are not using any CL strategy, forgetting will occur in class-incremental settings. This is not a problem related to model adaptation.

**Comment by PabloMese** [comment]
But the 'forggeting' that I'm experiencing is not related to the CL strategy. Between incremental experiences when the model is adapted to include the new classes is when the 'forggeting ocuurs'

**Comment by AndreaCossu** [comment]
This is the expected behavior in class-incremental. Introduction of new classes causes forgetting unless a proper CL strategy is used.

**Comment by PabloMese** [comment]
Before training?

**Comment by HamedHemati** [comment]
@PabloMese are you saying that the accuracy of previous classes after only dynamically extending the linear layer (even before training on the new task) goes down?

**Comment by HamedHemati** [comment]
> @PabloMese are you saying that the accuracy of previous classes goes down after only dynamically extending the linear layer (even before training on the new task)?

If this is the case, then that could be possible because of the random initialization of the new node weights. Maybe initializing the new nodes with very small weights and zero bias can help?

**Comment by PabloMese** [comment]
> @PabloMese are you saying that the accuracy of previous classes goes down after only dynamically extending the linear layer (even before training on the new task)?

That's exactly my problem. Sorry if I have not explained myself well.

**Comment by PabloMese** [comment]
> > @PabloMese are you saying that the accuracy of previous classes goes down after only dynamically extending the linear layer (even before training on the new task)?
> 
> If this is the case, then that could be possible because of the random initialization of the new node weights. Maybe initializing the new nodes with very small weights and zero bias can help?

I tried to initialize them with zeros in weights and bias but this does not make sense because logits can be negative before the final softmax.

**Comment by HamedHemati** [comment]
> 
> I tried to initialize them with zeros in weights and bias but this does not make sense because logits can be negative before the final softmax.

If you're using cross-entropy, that's not a problem. It will automatically compute the softmax of the logits and everything will become positive.

**Comment by AndreaCossu** [comment]
If that's the case, changing the initialization as suggested by @HamedHemati should work. Can you provide a minimal working example to reproduce the behavior?

**Comment by PabloMese** [comment]
I'm collecting the necessary code and data to reproduce the behavior and I will share it asap.
